### ISSUE TITLE
too many jobstotal jobs = 100

### ISSUE THREAD

**Issue by G10h4ck** [pupnp/pupnp#178]
In RetroShare we use libupnp for getting external IP address and to forward ports on NAT, we often receive reports of the log being flooded by messages like this one

`too many jobstotal jobs = 100`

Finally it happened to me too so I was able to track down it came from libupnp

https://github.com/pupnp/pupnp/blob/master/upnp/src/threadutil/ThreadPool.c#L825

Moreover it seems that the more UPnP enabled devices are in the same network the more it is likely for this to happen, I suppose that sadly log flooding is not the only consequence.

So first of all an easy fix so it less likely to mangle other messages is to add a `\n` at the end of the message, and then it would be really helpful to print a stacktrace when this happens, so one can get an indication if this problem is triggered by library user code, you can find snippets of how to print a stacktrace in C++ easily on the internet, you can even adapt the code from RetroShare

https://github.com/RetroShare/RetroShare/blob/master/libretroshare/src/util/stacktrace.h

Cheers, and thanks for caring of libupnp ♡

**Comment by whyman** [comment]
First question of course is, which version of linupnp are you running, and can you confirm this behavour with the latest 1.12.1 version?

**Comment by G10h4ck** [comment]
It happened with a few versions of the 1.6.x series, now I am compiling with 1.12.1 from gentoo repository to see If I can reproduce the problem there, in any case I think the improvements on the message that could be eventually be printed should be done anyway, would you accept a pull request improving on that?

**Comment by whyman** [comment]
There was some discussion in 2017 about this issue: #29

**Comment by G10h4ck** [comment]
> There was some discussion in 2017 about this issue: #29

Yeah I read that before but seems it is not related to our case, I have been inspecting our code and it seems very unlikely that we are calling some Upnp* function in a loop.

**Comment by G10h4ck** [comment]
I can confirm this still happens with libupnp-1.12.1 after half an hour of running my log started being mangled with `total jobs = 100, too many jobstotal jobs = 100, too many jobstotal jobs = 100, too many jobs`

**Comment by G10h4ck** [comment]
I have opened a PR so at least the message is easier to track and avoid mangling other messages in the log https://github.com/pupnp/pupnp/pull/179

**Comment by mrjimenez** [comment]
Hi G10h4ck,

Can you try running configure with --disable-blocking_tcp_connections ?

Regards,
Marcelo.

**Comment by G10h4ck** [comment]
> Hi G10h4ck,
> 
> Can you try running configure with --disable-blocking_tcp_connections ?
> 
> Regards,
> Marcelo.

It is already compiled by default with that option on gentoo, and in installation it have this gentoo flags `Installed versions:  1.12.1(10:16:41 AM 06/03/2020)(ipv6 reuseaddr -blocking-tcp -debug -doc -samples -static-libs)`

In particular `-blocking-tcp` means `--disable-blocking_tcp_connections`

**Comment by mrjimenez** [comment]
Hi @G10h4ck,

The issues you reported have now been addressed — here is a summary of what was done:

**Already merged to `main` (before this PR):**
- PR #179 (yours!) fixed the message format: added the `libupnp ThreadPoolAdd` prefix and the missing newline that was causing log mangling.
- A follow-up commit added spam suppression: the "too many jobs" message is now printed only **once** (on the first drop); when the queue recovers, a single "queue available again, N job(s) dropped" line is emitted instead of repeated flooding.

**Now in PR #554:**
1. **Backtrace on first overflow** — on Linux and macOS (where `execinfo.h` is available), `backtrace_symbols_fd()` is called immediately after the first "too many jobs" message, so you can see exactly which call path is filling the queue. This is the diagnostic you asked for. A new `UPNP_ENABLE_BACKTRACE` CMake option / `--enable-backtrace` configure switch controls it (enabled by default).

2. **Raised `DEFAULT_MAX_JOBS_TOTAL` from 100 to 1000** — the old limit was easily hit on networks with many UPnP devices. The new default gives much more headroom while remaining trivially small in terms of memory. `UpnpSetMaxJobsTotal()` is still available if you need a different ceiling.

3. **Regression test** — a new test verifies the overflow detection counter and (on backtrace-capable platforms) that the backtrace output actually appears on stderr.

Closing this issue as fixed via PR #554. Please feel free to reopen if the problem persists or if you have further feedback.

Thank you for the detailed report and for PR #179! ♡
